staging/src/k8s.io/apiserver/pkg/util/trace/trace.go

```diff
@@ -48,22 +48,42 @@ func (t *Trace) Step(msg string) {
 }
 
 func (t *Trace) Log() {
+	// an explicit logging request should dump all the steps out at the higher level
+	t.logWithStepThreshold(0)
+}
+
+// logWithStepThreshold will log exceptionally long steps at a higher log level than all the other "normal" steps.
+// If the threshold is zero, then it just dumps at the higher level.
+func (t *Trace) logWithStepThreshold(stepThreshold time.Duration) {
 	endTime := time.Now()
-	var buffer bytes.Buffer
+	var allStepsBuffer bytes.Buffer
+	var importantStepsBuffer bytes.Buffer
 
-	buffer.WriteString(fmt.Sprintf("Trace %q (started %v):\n", t.name, t.startTime))
+	allStepsBuffer.WriteString(fmt.Sprintf("Trace %q (started %v):\n", t.name, t.startTime))
 	lastStepTime := t.startTime
 	for _, step := range t.steps {
-		buffer.WriteString(fmt.Sprintf("[%v] [%v] %v\n", step.stepTime.Sub(t.startTime), step.stepTime.Sub(lastStepTime), step.msg))
+		stepDuration := step.stepTime.Sub(lastStepTime)
+		allStepsBuffer.WriteString(fmt.Sprintf("[%v] [%v] %v\n", step.stepTime.Sub(t.startTime), stepDuration, step.msg))
+		if stepThreshold != 0 && stepDuration > stepThreshold {
+			importantStepsBuffer.WriteString(fmt.Sprintf("[%v] [%v] %v\n", step.stepTime.Sub(t.startTime), stepDuration, step.msg))
+		}
 		lastStepTime = step.stepTime
 	}
-	buffer.WriteString(fmt.Sprintf("%q [%v] [%v] END\n", t.name, endTime.Sub(t.startTime), endTime.Sub(lastStepTime)))
-	glog.Info(buffer.String())
+	allStepsBuffer.WriteString(fmt.Sprintf("%q [%v] [%v] END\n", t.name, endTime.Sub(t.startTime), endTime.Sub(lastStepTime)))
+	if stepThreshold == 0 {
+		glog.Info(allStepsBuffer.String())
+		return
+	}
+
+	glog.Info(importantStepsBuffer.String())
+	glog.V(4).Info(allStepsBuffer.String())
 }
 
 func (t *Trace) LogIfLong(threshold time.Duration) {
 	if time.Since(t.startTime) >= threshold {
-		t.Log()
+		// if any step took more than 10% of the total time, it deserves a higher log level
+		stepThreshold := threshold / 10
```

Comment:
I'm wondering if we should replace the `10` with something like `stepThreshold= threshold / (len(t.steps) + 1)`